### PR TITLE
[SOLUTION] Exercise 2

### PR DESCRIPTION
**Pull request by TheInkedEngineer**

### CODE REVIEW

File: Exercise-02/Exercise02/main.swift [hunk 1]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+  
+  for entry in tempEntries {
+    if abs(date.timeIntervalSince1970 - entry.date.timeIntervalSince1970) < minDifference {
+      throw TemperatureError.tooEarly
+    }
+  }
+  
+}
+
+func getTemperatureEntries(for city: CityName, historical: inout Historical) -> [TemperatureEntry]? {
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: You don't need to pass `historical` as inout parameter for read-only purposes.

File: Exercise-02/Exercise02/main.swift [hunk 2]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+  
+  for entry in historical {
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: `Historical` is a dictionary, by definition dicts provide fast access (O(1)) to their elements via key.
If you iterate through the dictionary pairs (O(n)), you will lose this huge benefit.

File: Exercise-02/Exercise02/main.swift [hunk 3]
```diff
@@ -1,38 +1,60 @@
...
   print("Welcome to the Temperature Management System (TMS)")
   print("Those are the operations you can perform: ")
-  //TODO: add commands of your operations
   
-  print("Q - To quit the application quit")
+  print("I - Insert a new temperature")
+  print("R - Remove a temperature")
+  print("P - Print all temperatures for a given city")
+  print("A - Average temperature for a given city between two dates")
+  print("M - Max and Min tempratures for a given city")
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: tempratures -> temperatures

File: Exercise-02/Exercise02/main.swift [hunk 4]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+    let temp = getTemperature(),
+    let unit = getMeasurementUnit()
+  else {
+    return
+  }
+  
+  var tempEntries: [TemperatureEntry] = []
+  
+  if
+    let historicalEntries = getTemperatureEntries(for: city, historical: &historical) {
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: You can use the nil-coalescing operator:
`var tempEntries = getTemperatureEntries(for: city, historical: historical) ?? []`

File: Exercise-02/Exercise02/main.swift [hunk 5]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+  do {
+    try ValidateTimeOfInsertion(at: date, in: tempEntries)
+  } catch TemperatureError.tooEarly {
+    print("You can only insert temperatures every four hours")
+    return
+  } catch {
+    print("Internal unkown error")
+    return
+  }
+
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: You can move the access to historical here, after all the needed checks have been made.

File: Exercise-02/Exercise02/main.swift [hunk 6]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+func handleDeleteData(in historical: inout Historical) {
+  
+  guard
+    let city = getCity(),
+    var tempEntries = getTemperatureEntries(for: city, historical: &historical),
+    let date = getDate()
+  
+  else {
+      return
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: Print something in order to explain to the user that the city is not in the db.

File: Exercise-02/Exercise02/main.swift [hunk 7]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+  let dateFormatter = DateFormatter()
+  dateFormatter.dateFormat = "dd/MM/yyyy' 'HH:mm"
+  
+  guard
+    let userInput = readLine(),
+    !userInput.isEmpty,
+    let date = dateFormatter.date(from: userInput)
+  
+  else {
+    printInvalidEntry(reason: "Date format should be: \(dateFormatter.dateFormat!)")
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: `That's what bangs are for.. Kill hope` cit. MB

File: Exercise-02/Exercise02/main.swift [hunk 8]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+  }
+  
+  return temp
+  
+}
+
+func validateTemperature(_ temp: Double, in unit: MeasurementSystem) throws{
+  
+  if
+    convertTemperature(temp, from: unit, to: .celcius) >= 50 {
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: You can convert the temperature just once.

File: Exercise-02/Exercise02/main.swift [hunk 9]
```diff
@@ -60,7 +97,305 @@ func runApplication() {
...
+  }
+  
+  if
+    convertTemperature(temp, from: unit, to: .celcius) <= -50 {
+    throw TemperatureError.tooLow
+  }
+  
+}
+
+func ValidateTimeOfInsertion(at date: Date, in tempEntries: [TemperatureEntry]) throws {
```
Comment (AlfioRussoBS) [inline Exercise-02/Exercise02/main.swift]: Validate -> validate

File: Exercise-02/Exercise02/main.swift [hunk 10]
```diff
@@ -60,7 +97,310 @@ func runApplication() {
...
+  case (.kelvin, .celsius):
+    return value - 273.15
+    
+  case (.kelvin, .fahrenheit):
+    return value * 1.8 - 459.67
+  
+  default:
+    // should not happen
+    // workaround for exhaustive search
+    // already checking from != to with guard above
```
Comment (TheInkedEngineer) [inline Exercise-02/Exercise02/main.swift]: @AlfioRussoBS intendo qui

**Pull request merged**